### PR TITLE
[Oneway][Part 3] Encoding layers can handle oneway requests and registration

### PR DESCRIPTION
**Pull request by AlexAPB**
Encoding can register oneway procedures and handle oneway requests

### CODE REVIEW

File: encoding/json/register.go [hunk 1]
```diff
@@ -66,13 +66,38 @@ func Procedure(name string, handler interface{}) []transport.Registrant {
 	}
 }
 
+// OnewayProcedure builds a Registrant from the given JSON handler. handler must be
+// a function with a signature similar to,
+//
+// 	f(ctx context.Context, reqMeta yarpc.ReqMeta, body $reqBody) error
```
Comment (abhinav) [inline encoding/json/register.go]: Can you also add something to the module docs? (json/doc.go)

File: encoding/raw/inbound.go [hunk 2]
```diff
@@ -29,9 +29,10 @@ import (
 	"go.uber.org/yarpc/transport"
 )
 
-// rawHandler adapts a Handler into a transport.UnaryHandler
+// rawHandler adapts a Handler into a transport.UnaryHandler and transport.OnewayHandler
 type rawHandler struct {
-	UnaryHandler UnaryHandler
+	UnaryHandler  UnaryHandler
+	OnewayHandler OnewayHandler
```
Comment (abhinav) [inline encoding/raw/inbound.go]: I would define two separate types here: rawUnaryHandler and rawOnewayHandler

File: encoding/raw/inbound.go [hunk 3]
```diff
@@ -60,3 +61,22 @@ func (r rawHandler) Handle(ctx context.Context, treq *transport.Request, rw tran
...
+	reqBody, err := ioutil.ReadAll(treq.Body)
+	if err != nil {
+		return err
+	}
+
+	reqMeta := meta.FromTransportRequest(treq)
+	err = r.OnewayHandler(ctx, reqMeta, reqBody)
+	if err != nil {
+		return err
+	}
```
Comment (abhinav) [inline encoding/raw/inbound.go]: Can shorten to `return r.OnewayHandler(ctx, reqMeta, reqBody)`

File: encoding/raw/register.go [hunk 4]
```diff
@@ -47,7 +47,23 @@ func Procedure(name string, handler UnaryHandler) []transport.Registrant {
...
+type OnewayHandler func(context.Context, yarpc.ReqMeta, []byte) error
+
+// OnewayProcedure builds a Registrant from the given raw handler
+func OnewayProcedure(name string, handler OnewayHandler) []transport.Registrant {
+	return []transport.Registrant{
+		{
+			Procedure: name,
+			HandlerSpec: transport.HandlerSpec{
+				Type:          transport.Oneway,
+				OnewayHandler: rawHandler{OnewayHandler: handler},
```
Comment (abhinav) [inline encoding/raw/register.go]: Declare and use `rawOnewayHandler` here

File: encoding/thrift/inbound.go [hunk 5]
```diff
@@ -33,11 +33,13 @@ import (
...
 
-// thriftHandler wraps a Thrift Handler into a transport.UnaryHandler
+// thriftHandler wraps a Thrift Handler into a transport.UnaryHandler and
+// transport.OnewayHandler
 type thriftHandler struct {
-	UnaryHandler UnaryHandler
-	Protocol     protocol.Protocol
-	Enveloping   bool
+	UnaryHandler  UnaryHandler
+	OnewayHandler OnewayHandler
```
Comment (abhinav) [inline encoding/thrift/inbound.go]: Same as above. Separate types.

**Comment by AlexAPB** [comment]
updated

File: encoding/raw/inbound.go [hunk 6]
```diff
@@ -29,12 +29,13 @@ import (
...
 
-// rawHandler adapts a Handler into a transport.UnaryHandler
-type rawHandler struct {
-	UnaryHandler UnaryHandler
-}
+// rawUnaryHandler adapts a Handler into a transport.UnaryHandler
+type rawUnaryHandler struct{ UnaryHandler }
+
+// rawOnewayHandler adapts a Handler into a transport.OnewayHandler
+type rawOnewayHandler struct{ OnewayHandler OnewayHandler }
```
Comment (abhinav) [inline encoding/raw/inbound.go]: You don't need to provide the field name if you're embedding like the other Handler.

**Pull request merged**